pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py

```diff
@@ -218,6 +223,45 @@ def _request_generator_thread(self):
             _LOGGER.debug('Sending request:\n%r', request)
             yield request
 
+    def _stop_request_generator(self, request_generator):
+        """Ensure a request generator is closed.
+
+        This **must** be done when recovering from a retry-able exception.
+        If not, then an inactive request generator (i.e. not attached to any
+        actual RPC) will be trying to access the same request queue as the
+        active request generator.
+
+        In addition, we want the gRPC thread consuming to cleanly exit so
+        that system resources are not wasted.
+
+        Args:
+            request_generator (Generator): A streaming pull request generator
+                returned from :meth:`_request_generator_thread`.
+
+        Raises:
+            ValueError: If ``.close()`` fails for any reason other than
+                "generator already executing".
+            ValueError: If the queue is not empty and the generator is
+                running.
+        """
+        with self._put_lock:
+            try:
+                request_generator.close()
```

Comment:
There **might** is a seemingly impossible-to-trigger corner case where the generator is **still running** for the few nanoseconds **after** the blocking `.get()` returns. In that case, we'd either need to

- Somehow force the gRPC thread that is consuming the request generator to knock it off
- Just try to `.close()` one more time if we detect the queue isn't empty

----

FWIW I don't think it's worth accounting for that impossible-to-trigger corner case.